### PR TITLE
drawable blocks

### PR DESCRIPTION
**Pull request by kecajjo**

### CODE REVIEW

File: bl/drawable/inc/drawable.hpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    virtual char getColor() const = 0;
+    virtual std::list<std::pair<int, int>> getShape() const = 0;
+
+    /*!
+        Returns list of all children of the object that can be visualised
+        tuple: position_x, position_y, ptr_to_child_element
+        TODO how to make this return list of either weak or shared ptr to any class derived from
+       drawable
+    */
+    virtual std::list<DrawablePosition<std::shared_ptr<Drawable>>> getChildDrawables() const {
```
Comment (ProrokWielki) [inline bl/drawable/inc/drawable.hpp]: why is this virtual, why would you like to override it?

File: bl/drawable/inc/drawable.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+   public:
+    // TODO change to shared_ptr only???
+    template <typename T, std::enable_if_t<is_smart_ptr<T>::value, bool> = true,
+              std::enable_if_t<
+                  std::is_base_of<Drawable, std::decay_t<decltype(*std::declval<T>())>>::value,
+                  bool> = true>
+    struct DrawablePosition {
+        int x;
+        int y;
+        T objHdl;
```
Comment (ProrokWielki) [inline bl/drawable/inc/drawable.hpp]: Why is that needed?

File: bl/drawable/inc/drawable.hpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+#include <list>
+#include <memory>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+class Drawable {
+   private:
+    template <typename T>
+    struct is_smart_ptr : std::false_type {};
```
Comment (ProrokWielki) [inline bl/drawable/inc/drawable.hpp]: i dont knwo if you really need it, but if you do, i would put it in some util header.

File: bl/drawable/inc/drawable.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                  std::is_base_of<Drawable, std::decay_t<decltype(*std::declval<T>())>>::value,
+                  bool> = true>
+    struct DrawablePosition {
+        int x;
+        int y;
+        T objHdl;
+    };
+
+    virtual ~Drawable() = default;
+    virtual char getColor() const = 0;
```
Comment (ProrokWielki) [inline bl/drawable/inc/drawable.hpp]: How about creating seperate class/struct for pixel type, it could hold rgb values, if ever decide to use it with something which can support colors your code will support it.  And for the drowing angine whcih does not support colors it can any way take pixels as a data, but convert it to chars by it self.

File: bl/board/inc/board.hpp [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+class Board {
+   private:
+    template <typename>
+    struct is_tuple : std::false_type {};
+    template <typename pos, typename ptr>
+    struct is_tuple<std::tuple<pos, pos, ptr>> : std::true_type {};
+
+   public:
+    template <
+        class... Args, std::enable_if_t<(is_tuple<Args>::value && ...), bool> = true,
```
Comment (ProrokWielki) [inline bl/board/inc/board.hpp]: Very nice template magic, but its is completely unnecessary over complicated. You could just do
addBlocks(Block* ... block)

File: bl/blocks/src/tshape.cpp [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#include "tshape.hpp"
+
+std::list<std::pair<int, int>> TShape::getShape() const {
```
Comment (ProrokWielki) [inline bl/blocks/src/tshape.cpp]: If you'd put the list as the class member and initlize it in constructor and here just return it here it would be much faster

File: bl/blocks/src/square.cpp [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#include "square.hpp"
+
+std::list<std::pair<int, int>> Square::getShape() const {
+    std::list<std::pair<int, int>> l;
+    l.push_back(std::pair<int, int>(0, 0));
```
Comment (ProrokWielki) [inline bl/blocks/src/square.cpp]: If you'd put the list as the class member and initlize it in constructor and here just return it here it would be much faster

File: bl/blocks/src/lshape.cpp [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#include "lshape.hpp"
+
+std::list<std::pair<int, int>> LShape::getShape() const {
+    std::list<std::pair<int, int>> l;
+    l.push_back(std::pair<int, int>(0, 0));
```
Comment (ProrokWielki) [inline bl/blocks/src/lshape.cpp]: If you'd put the list as the class member and initlize it in constructor and here just return it here it would be much faster

File: bl/blocks/src/diagonal.cpp [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#include "diagonal.hpp"
+
+std::list<std::pair<int, int>> Diagonal::getShape() const {
+    std::list<std::pair<int, int>> l;
+    l.push_back(std::pair<int, int>(0, 0));
```
Comment (ProrokWielki) [inline bl/blocks/src/diagonal.cpp]: If you'd put the list as the class member and initlize it in constructor and here just return it here it would be much faster

File: bl/blocks/inc/tshape.hpp [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+#include <list>
+#include <utility>
+
+#include "block.hpp"
+
+class TShape : public Block {
+   public:
+    explicit TShape() {}
+    std::list<std::pair<int, int>> getShape() const override;
+    ~TShape() = default;
```
Comment (ProrokWielki) [inline bl/blocks/inc/tshape.hpp]: add "override"

File: bl/blocks/inc/square.hpp [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+#include <list>
+#include <utility>
+
+#include "block.hpp"
+
+class Square : public Block {
+   public:
+    explicit Square() {}
+    std::list<std::pair<int, int>> getShape() const override;
+    ~Square() = default;
```
Comment (ProrokWielki) [inline bl/blocks/inc/square.hpp]: add "override"

File: bl/blocks/inc/lshape.hpp [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+#include <list>
+#include <utility>
+
+#include "block.hpp"
+
+class LShape : public Block {
+   public:
+    explicit LShape() {}
+    std::list<std::pair<int, int>> getShape() const override;
+    ~LShape() = default;
```
Comment (ProrokWielki) [inline bl/blocks/inc/lshape.hpp]: add "override"

File: bl/blocks/inc/diagonal.hpp [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+#include <list>
+#include <utility>
+
+#include "block.hpp"
+
+class Diagonal : public Block {
+   public:
+    explicit Diagonal() {}
+    std::list<std::pair<int, int>> getShape() const override;
+    ~Diagonal() = default;
```
Comment (ProrokWielki) [inline bl/blocks/inc/diagonal.hpp]: add "override"

File: bl/blocks/inc/block.hpp [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+                                       // the same as base one
+
+    // move constructor and move assignment operator not needed, because only
+    // static variables
+    // explicit Block(Block &&b);
+    // Block &operator=(Block &&b);
+    char getColor() const override { return color; }
+
+    // virtual std::list<std::pair<int, int>> getShape() const  = 0;
+    virtual ~Block() = default;
```
Comment (ProrokWielki) [inline bl/blocks/inc/block.hpp]: it should be marked override not virtual

File: bl/drawable/inc/drawable.hpp (hunk 1 above)
Comment (kecajjo) [reply bl/drawable/inc/drawable.hpp]: If object holds 2 types of drawables which each needs to be on a separate list due to their different functionality in the backend (for example board having trees and tanks, sometimes you would need to iterate through all the tanks, but dont want to iterate through trees) it would need to have 2 ptrs to an object (in list of drawables and in list of tanks/trees). If you overload this function then you dont need to have a list of drawables, you can just return list that is sum of lists of tanks and trees inside it. So it would be up to the class to know which of its children is a drawable and return it (it would need to know if its drawable anyway to add it to the list)

Let me know if this way of thinking is correct

File: bl/board/inc/board.hpp (hunk 5 above)
Comment (kecajjo) [reply bl/board/inc/board.hpp]: if i do addBlocks(Block*  block...) I cannot use fold expression and have to use va_start, va_end and va_arg and do it C-style way. Is it better?

File: bl/drawable/inc/drawable.hpp (hunk 2 above)
Comment (kecajjo) [reply bl/drawable/inc/drawable.hpp]: Not needed, fixed

File: bl/drawable/inc/drawable.hpp (hunk 3 above)
Comment (kecajjo) [reply bl/drawable/inc/drawable.hpp]: not needed anymore, removed

File: bl/drawable/inc/drawable.hpp (hunk 4 above)
Comment (kecajjo) [reply bl/drawable/inc/drawable.hpp]: fixed

File: bl/blocks/src/tshape.cpp (hunk 6 above)
Comment (kecajjo) [reply bl/blocks/src/tshape.cpp]: fixed

File: bl/blocks/src/square.cpp (hunk 7 above)
Comment (kecajjo) [reply bl/blocks/src/square.cpp]: fixed

File: bl/blocks/src/lshape.cpp (hunk 8 above)
Comment (kecajjo) [reply bl/blocks/src/lshape.cpp]: fixed

File: bl/blocks/src/diagonal.cpp (hunk 9 above)
Comment (kecajjo) [reply bl/blocks/src/diagonal.cpp]: fixed

File: bl/blocks/inc/tshape.hpp (hunk 10 above)
Comment (kecajjo) [reply bl/blocks/inc/tshape.hpp]: fixed

File: bl/blocks/inc/square.hpp (hunk 11 above)
Comment (kecajjo) [reply bl/blocks/inc/square.hpp]: fixed

File: bl/blocks/inc/lshape.hpp (hunk 12 above)
Comment (kecajjo) [reply bl/blocks/inc/lshape.hpp]: fixed

File: bl/blocks/inc/diagonal.hpp (hunk 13 above)
Comment (kecajjo) [reply bl/blocks/inc/diagonal.hpp]: fixed

File: bl/blocks/inc/block.hpp (hunk 14 above)
Comment (kecajjo) [reply bl/blocks/inc/block.hpp]: fixed

File: bl/board/inc/board.hpp (hunk 5 above)
Comment (ProrokWielki) [reply bl/board/inc/board.hpp]: ok, good poitnt, so lets do it like that 

```cpp
template<typename ... Args>
void add_block<tuple<Block*,int,int> block, Args... blocks>
{
blocks.push_back( {Drawable::Position{std::get<0>(blck), std::get<1>(blck)}, std::get<2>(blck)}),
 add_block(block...)
}
```

File: bl/drawable/inc/drawable.hpp (hunk 1 above)
Comment (ProrokWielki) [reply bl/drawable/inc/drawable.hpp]: I don't get the usecase to be honest :D let's discuss it in person :D

File: bl/board/inc/board.hpp (hunk 5 above)
Comment (ProrokWielki) [reply bl/board/inc/board.hpp]: Oh, you can also just do 
```cpp
void add_block(std::initializer_list<tuple<Block*,int,int>> blocks) 
{

}
```
And you don't have to worry about passed types :)

File: bl/drawable/inc/drawable.hpp (hunk 1 above)
Comment (kecajjo) [reply bl/drawable/inc/drawable.hpp]: fixed

File: bl/board/inc/board.hpp (hunk 5 above)
Comment (kecajjo) [reply bl/board/inc/board.hpp]: Automatic template parameter deduction does not work, I would need to explicitly say what parameters are given (i.e std::tuple<int, int, std::shared_ptr<Block>>{1, 1, p} instead of {1, 1, p}) so I changed this function to add just 1 block :(

Comment (ProrokWielki) [reply bl/board/inc/board.hpp]: ok, if you say so :) lets leave it as is

**Review by ProrokWielki — APPROVED** [review]

File: bl/board/inc/board.hpp (hunk 5 above)
Comment (kecajjo) [reply bl/board/inc/board.hpp]: didnt expect you checking it at this hour so i wrote comments before pushing changes :P my bad

File: bl/blocks/src/block.cpp
Comment (ProrokWielki) [inline bl/blocks/src/block.cpp:1]: why?

File: bl/board/src/board.cpp
Comment (ProrokWielki) [inline bl/board/src/board.cpp:1]: why?

File: bl/board/inc/board.hpp [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+#include <type_traits>
+#include <unordered_set>
+
+#include "block.hpp"
+#include "drawable.hpp"
+
+// TODO rest of the class
+class Board : public Drawable {
+   private:
+    template <typename>
```
Comment (ProrokWielki) [inline bl/board/inc/board.hpp:12]: do you still need those?
